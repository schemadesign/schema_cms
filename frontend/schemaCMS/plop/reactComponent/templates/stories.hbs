import React from 'react';
import { storiesOf } from '@storybook/react';

import { {{ pascalCase name }} } from './{{ camelCase name }}.component';

export const defaultProps = {};

storiesOf('{{ pascalCase name }}', module).add('Default', () => <{{ pascalCase name }} {...defaultProps} />);
