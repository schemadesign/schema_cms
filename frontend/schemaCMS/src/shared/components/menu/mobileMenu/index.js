export { MobileMenu } from './mobileMenu.component';
