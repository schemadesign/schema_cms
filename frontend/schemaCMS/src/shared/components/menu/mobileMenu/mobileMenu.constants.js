export const LINK_ITEM = 'link_item';
export const DIVIDER = 'divider';
