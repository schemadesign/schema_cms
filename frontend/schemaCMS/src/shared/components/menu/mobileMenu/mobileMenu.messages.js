/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  users: {
    id: 'shared.components.mobileMenu.users',
    defaultMessage: 'Users',
  },
  about: {
    id: 'shared.components.mobileMenu.about',
    defaultMessage: 'About Schema CMS',
  },
  api: {
    id: 'shared.components.mobileMenu.api',
    defaultMessage: 'API Doc',
  },
  repository: {
    id: 'shared.components.mobileMenu.repository',
    defaultMessage: 'GitHub Repository',
  },
  projects: {
    id: 'shared.components.mobileMenu.projects',
    defaultMessage: 'Projects',
  },
  logOut: {
    id: 'shared.components.mobileMenu.logOut',
    defaultMessage: 'Log out',
  },
  settings: {
    id: 'shared.components.mobileMenu.settings',
    defaultMessage: 'Settings',
  },
});
