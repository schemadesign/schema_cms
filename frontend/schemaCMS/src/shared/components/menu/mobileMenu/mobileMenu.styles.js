import styled, { css } from 'styled-components';
import { Typography } from 'schemaUI';
import { Link } from 'react-router-dom';
import { ifElse, equals } from 'ramda';

import { media } from '../../../../theme/media';

const HEADER_HEIGHT = 54;
const CONTENT_PADDING = 40;

const getItemTextColor = ({ active, theme }) =>
  ifElse(
    equals(true),
    () => css`
      color: ${theme.text};
    `,
    () => css`
      color: ${theme.secondaryText};
    `
  )(active);

export const Container = styled.div`
  font-weight: normal;
  width: calc(100% - 50px);

  ${media.desktop`
    display: none;
  `};
`;

export const HeaderWrapper = styled.div``;

export const MenuHeader = styled.div`
  height: ${HEADER_HEIGHT}px;
`;

export const Content = styled.div`
  overflow: auto;
  max-height: calc(100vh - 134px);
  padding: ${CONTENT_PADDING}px 0;
  -webkit-overflow-scrolling: touch;

  ${media.desktop`
    padding-bottom: 20px;
    max-height: calc(100vh - ${2 * CONTENT_PADDING + HEADER_HEIGHT}px);
  `};
`;

export const List = styled.div`
  list-style: none;
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
`;

export const Item = styled.div`
  line-height: 1.5;
  margin: 2px 0;
  cursor: pointer;
  font-size: 24px;
  border-bottom: 2.4px solid ${({ theme, active }) => (active ? theme.header.text : theme.header.border)};
  padding: 8px 0 13px;

  a,
  div {
    text-decoration: none;
    ${getItemTextColor};
    display: block;
  }
`;

export const HelperList = styled.div`
  margin: 10px 0;
  display: flex;
  flex-direction: column;
`;

export const HelperLink = styled(Link)`
  color: ${({ theme }) => theme.text};
  font-size: 14px;
  text-decoration: none;
  margin: 12px 0;
`;

export const Title = styled(Typography.H2)`
  word-break: break-word;
`;

export const Subtitle = styled(Typography.H1)`
  word-break: break-word;
`;

export const menuStyles = {
  zIndex: 99999,
  maxWidth: null,
  position: 'absolute',
};

export const closeButtonStyles = {
  backgroundColor: 'transparent',
};

export const customButtonStyles = {
  right: -6,
};

export const Divider = styled.div``;
