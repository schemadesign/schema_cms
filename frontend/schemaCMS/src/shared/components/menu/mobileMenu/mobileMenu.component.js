import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Header, Menu } from 'schemaUI';
import { always, ifElse, cond, propEq, T, isEmpty } from 'ramda';
import { FormattedMessage } from 'react-intl';

import {
  closeButtonStyles,
  Container,
  Content,
  HeaderWrapper,
  MenuHeader,
  menuStyles,
  Item,
  Subtitle,
  Title,
  List,
  Divider,
  HelperLink,
  HelperList,
  customButtonStyles,
} from './mobileMenu.styles';
import messages from './mobileMenu.messages';
import { LogoutModal } from '../../logoutModal';
import { DIVIDER, LINK_ITEM, HELPER_LINK, SETTINGS_ID } from './mobileMenu.constants';
import { handleToggleMenu } from '../../../utils/helpers';

const HELPER_LINKS = [
  { label: <FormattedMessage {...messages.about} />, to: '/', id: 'aboutNavBtn', type: HELPER_LINK },
  { label: <FormattedMessage {...messages.api} />, to: '/', id: 'apiNavBtn', type: HELPER_LINK },
  { label: <FormattedMessage {...messages.repository} />, to: '/', id: 'repositoryNavBtn', type: HELPER_LINK },
];

export class MobileMenu extends PureComponent {
  static propTypes = {
    headerTitle: PropTypes.node,
    headerSubtitle: PropTypes.node,
    options: PropTypes.array,
    iconComponent: PropTypes.element,
    active: PropTypes.string,
  };

  static defaultProps = {
    options: [],
  };

  state = {
    isMenuOpen: false,
    logoutModalOpen: false,
  };

  componentWillUnmount() {
    document.documentElement.classList.remove('hideScroll');
  }

  handleToggleMenu = () => handleToggleMenu(this);

  handleLogout = () => {
    this.setState({
      logoutModalOpen: true,
    });
  };

  handleCancelLogout = () => {
    this.setState({
      logoutModalOpen: false,
    });
  };

  renderHeader = (title, subtitle) => (
    <HeaderWrapper>
      <Title id="headerTitle">{title}</Title>
      <Subtitle id="headerSubtitle">{subtitle}</Subtitle>
    </HeaderWrapper>
  );

  renderItem = ({ id, to = '', onClick = this.handleToggleMenu, label }, index) => (
    <Item key={index} active={id === this.props.active}>
      {ifElse(
        isEmpty,
        always(<div onClick={onClick}>{label}</div>),
        always(
          <Link id={id} to={to} onClick={onClick}>
            {label}
          </Link>
        )
      )(to)}
    </Item>
  );

  renderHelperLink = ({ id, to, label }, index) => (
    <HelperLink key={index} id={id} to={to} onClick={this.handleToggleMenu}>
      {label}
    </HelperLink>
  );

  renderOptions = (option, index) =>
    cond([
      [propEq('type', LINK_ITEM), () => this.renderItem(option, index)],
      [propEq('type', HELPER_LINK), () => this.renderHelperLink(option, index)],
      [propEq('type', DIVIDER), () => always(<Divider />)],
      [T, () => this.renderItem(option, index)],
    ])(option);

  render() {
    const { headerTitle, headerSubtitle, options, iconComponent } = this.props;
    const { logoutModalOpen, isMenuOpen } = this.state;
    const headerContent = this.renderHeader(headerTitle, headerSubtitle);

    const buttonProps = {
      onClick: this.handleToggleMenu,
      id: 'topHeaderOpenMenuBtn',
      type: 'button',
    };

    const closeButtonProps = {
      customStyles: closeButtonStyles,
      id: 'topHeaderCloseMenuButton',
      type: 'button',
    };

    const fixedMenuItems = [
      {
        label: <FormattedMessage {...messages.settings} />,
        to: '/settings',
        id: SETTINGS_ID,
        type: LINK_ITEM,
        page: 'settings',
      },
      {
        label: <FormattedMessage {...messages.logOut} />,
        onClick: this.handleLogout,
        type: LINK_ITEM,
        id: 'logoutNavBtn',
      },
    ];

    return (
      <Container>
        <Header buttonProps={buttonProps} customButtonStyles={customButtonStyles} iconComponent={iconComponent}>
          {headerContent}
        </Header>
        <Menu
          open={isMenuOpen}
          onClose={this.handleToggleMenu}
          customStyles={menuStyles}
          closeButtonProps={closeButtonProps}
        >
          <Content>
            <MenuHeader>{headerContent}</MenuHeader>
            <List>{options.map(this.renderOptions)}</List>
            <List>{fixedMenuItems.map(this.renderOptions)}</List>
            <HelperList>{HELPER_LINKS.map(this.renderOptions)}</HelperList>
          </Content>
        </Menu>
        <LogoutModal logoutModalOpen={logoutModalOpen} onAction={this.handleCancelLogout} redirectUrl="/logout" />
      </Container>
    );
  }
}
