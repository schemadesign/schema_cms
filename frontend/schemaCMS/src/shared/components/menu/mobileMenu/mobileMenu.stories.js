import React from 'react';
import { storiesOf } from '@storybook/react';

import { MobileMenu } from './mobileMenu.component';

const defaultProps = {};

storiesOf('MobileMenu', module).add('Default', () => <MobileMenu {...defaultProps} />);
