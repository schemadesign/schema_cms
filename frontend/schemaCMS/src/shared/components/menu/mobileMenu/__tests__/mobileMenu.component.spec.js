import React from 'react';
import { shallow } from 'enzyme';

import { MobileMenu } from '../mobileMenu.component';
import { defaultProps } from '../mobileMenu.stories';

describe('MobileMenu: Component', () => {
  const component = props => <MobileMenu {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render correctly', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });

  it('should show logout popup and hide menu', () => {
    const wrapper = render();
    wrapper.instance().handleToggleMenu();

    expect(wrapper.state().isMenuOpen).toEqual(true);

    wrapper.find('#logoutNavBtn').simulate('click');

    expect(wrapper.state().isMenuOpen).toEqual(false);
    expect(wrapper.state().logoutModalOpen).toEqual(true);
  });
});
