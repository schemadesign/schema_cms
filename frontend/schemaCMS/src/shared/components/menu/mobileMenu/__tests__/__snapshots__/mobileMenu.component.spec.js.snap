// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileMenu: Component should render correctly 1`] = `
<styled.div>
  <Component
    buttonProps={
      Object {
        "id": "topHeaderOpenMenuBtn",
        "onClick": [Function],
      }
    }
  >
    <styled.div>
      <Styled(H2)
        id="headerTitle"
      >
        <div>
          Header Title
        </div>
      </Styled(H2)>
      <Styled(H1)
        id="headerSubtitle"
      >
        <div>
          Header SubTitle
        </div>
      </Styled(H1)>
    </styled.div>
  </Component>
  <Component
    closeButtonProps={
      Object {
        "customStyles": Object {
          "backgroundColor": "transparent",
        },
        "id": "topHeaderCloseMenuButton",
      }
    }
    customStyles={
      Object {
        "maxWidth": null,
        "zIndex": 9999,
      }
    }
    onClose={[Function]}
    open={false}
  >
    <styled.div>
      <styled.div>
        <Styled(H2)
          id="headerTitle"
        >
          <div>
            Header Title
          </div>
        </Styled(H2)>
        <Styled(H1)
          id="headerSubtitle"
        >
          <div>
            Header SubTitle
          </div>
        </Styled(H1)>
      </styled.div>
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div
          active={false}
          key="0"
        >
          <Link
            id="1"
            onClick={[Function]}
            to="/sample/"
          >
            A label
          </Link>
        </styled.div>
        <styled.div
          active={false}
          key="1"
        >
          <Link
            id="2"
            onClick={[Function]}
            to="/sample/url"
          >
            Another label
          </Link>
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div
          active={false}
          key="0"
        >
          <Link
            id="settingsNavBtn"
            onClick={[Function]}
            to="/settings"
          >
            <FormattedMessage
              defaultMessage="Settings"
              id="shared.components.mobileMenu.settings"
              values={Object {}}
            />
          </Link>
        </styled.div>
        <styled.div
          active={false}
          key="1"
        >
          <div
            onClick={[Function]}
          >
            <FormattedMessage
              defaultMessage="Log out"
              id="shared.components.mobileMenu.logOut"
              values={Object {}}
            />
          </div>
        </styled.div>
      </styled.div>
      <styled.div>
        <Styled(Link)
          id="aboutNavBtn"
          key="0"
          onClick={[Function]}
          to="/"
        >
          <FormattedMessage
            defaultMessage="About Schema CMS"
            id="shared.components.mobileMenu.about"
            values={Object {}}
          />
        </Styled(Link)>
        <Styled(Link)
          id="apiNavBtn"
          key="1"
          onClick={[Function]}
          to="/"
        >
          <FormattedMessage
            defaultMessage="API Doc"
            id="shared.components.mobileMenu.api"
            values={Object {}}
          />
        </Styled(Link)>
        <Styled(Link)
          id="repositoryNavBtn"
          key="2"
          onClick={[Function]}
          to="/"
        >
          <FormattedMessage
            defaultMessage="GitHub Repository"
            id="shared.components.mobileMenu.repository"
            values={Object {}}
          />
        </Styled(Link)>
      </styled.div>
    </styled.div>
  </Component>
  <LogoutModal
    logoutModalOpen={false}
    onAction={[Function]}
    redirectUrl="/logout"
  />
</styled.div>
`;
